File: apps/mobile/src/core/notifications.ts
/**
 * Notifications (Phase 5c) — app-local, store-driven attention events.
 *
 * Design rules (memory: yenn-android-notifications-plan — NEVER event-sniff):
 * - Every event here is derived from state the phone core ALREADY ingested
 *   through the typed protocol (live output entries, session-failed messages,
 *   DM store inserts). No relay-event sniffing, no kind-existence checks.
 * - The decision is a PURE function of (event, app visibility): a foregrounded
 *   app never notifies — the UI itself is the notification. Per-conversation
 *   unread is respected upstream: the DM store only reports messages that
 *   count unread (active conversation suppresses them).
 * - Sync catch-up entries never notify: only the LIVE output path classifies
 *   entries, so replaying history after a reconnect can't cause a storm.
 * - The OS side is a port (`Notifier` in ports.ts): Tauri's notification
 *   plugin in production, a spy in tests, absent in the headless core.
 */
import type { OutputEntry } from '@codedeck/protocol';
import type { Notifier } from './ports';
import { sessionKeyOf } from './stores/ui';

// --- Event vocabulary ---

export type NotifyEvent =
  | { type: 'permission-request'; machine: string; sessionId: string; toolName?: string }
  | { type: 'question'; machine: string; sessionId: string }
  | { type: 'plan-approval'; machine: string; sessionId: string }
  | { type: 'session-finished'; machine: string; sessionId: string }
  | { type: 'session-failed'; machine: string; sessionId: string; reason?: string }
  | { type: 'dm-received'; peer: string; peerLabel?: string; preview?: string };

/** Cooldown/dedup scope: same key + type within the window → one notification. */
export function notifyKey(event: NotifyEvent): string {
  return event.type === 'dm-received'
    ? `dm ${event.peer}`
    : `${event.type} ${event.machine} ${event.sessionId}`;
}

// --- Cancellation tags (CDX-026c) ---
// Coarser than notifyKey on purpose: opening a session should clear EVERY
// delivered notification for it (permission answered, question read, turn
// finished), and opening a DM conversation clears that peer's.

/** Cancellation scope for all of one session's notifications. */
export const sessionNotifyTag = (machine: string, sessionId: string): string =>
  `session ${sessionKeyOf(machine, sessionId)}`;

/** Cancellation scope for one DM peer's notifications. */
export const dmNotifyTag = (peer: string): string => `dm ${peer}`;

/** The tag a given event's delivery is filed under. */
export function notifyTag(event: NotifyEvent): string {
  return event.type === 'dm-received'
    ? dmNotifyTag(event.peer)
    : sessionNotifyTag(event.machine, event.sessionId);
}

// --- The pure decision: event × app-visibility → notify or not ---

/**
 * A visible (foregrounded) app never posts OS notifications — the user is
 * looking at the UI. Everything that reaches this function already passed its
 * store-level gate (DM unread counting, live-output-only classification), so
 * hidden → notify, for every event type.
 */
export function decideNotify(_event: NotifyEvent, visible: boolean): boolean {
  return !visible;
}

/**
 * The in-app attention chime's pure decision (old-app `notifyIfNeeded` gate):
 * ping when the app is hidden OR the user is viewing a DIFFERENT session than
 * the event's — a visible user watching exactly this session needs no chime,
 * the card on screen IS the signal.
 *
 * DM events carry no session key and already passed the per-conversation
 * unread gate upstream (an open conversation's messages are never emitted at
 * all) — so any DM event that reaches this decision is by construction
 * "not currently being viewed" and pings. This mirrors how decideNotify sees
 * DM events: the store-level gate did the viewing check.
 *
 * `activeSessionKey` is sessionKeyOf(machine, sessionId) of the session the
 * user is looking at, or null when no session panel is in view.
 */
export function decidePing(
  event: NotifyEvent,
  visible: boolean,
  activeSessionKey: string | null,
): boolean {
  if (!visible) return true;
  if (event.type === 'dm-received') return true;
  return sessionKeyOf(event.machine, event.sessionId) !== activeSessionKey;
}

// --- Formatting ---

export interface NotificationContent {
  title: string;
  body: string;
}

export function formatNotifyEvent(event: NotifyEvent): NotificationContent {
  switch (event.type) {
    case 'permission-request':
      return {
        title: 'Permission needed',
        body: event.toolName
          ? `Claude wants to use ${event.toolName}`
          : 'Claude needs permission to proceed',
      };
    case 'question':
      return { title: 'Question from Claude', body: 'Claude is asking you a question' };
    case 'plan-approval':
      return { title: 'Plan ready for review', body: 'A plan is waiting for your approval' };
    case 'session-finished':
      return { title: 'Session finished', body: 'Claude finished the task' };
    case 'session-failed':
      return {
        title: 'Session failed',
        body: event.reason ?? 'The session ended with an error',
      };
    case 'dm-received':
      return {
        title: event.peerLabel ?? 'New message',
        body: event.preview ?? 'You received a direct message',
      };
    default: {
      const exhaustive: never = event;
      void exhaustive;
      return { title: 'CodeDeck', body: 'Attention needed' };
    }
  }
}

// --- Live-output entry classification (permission / question / plan / end) ---

/**
 * Map ONE live output entry to a notify event, or null. Mirrors the
 * displayEntries vocabulary (system+special cards, stream_end, error
 * specials) — the transcript renderer and the notifier must agree on what an
 * entry means.
 */
export function classifyOutputEntry(
  machine: string,
  sessionId: string,
  entry: OutputEntry,
): NotifyEvent | null {
  const special = entry.metadata?.special as string | undefined;
  if (entry.entryType === 'system') {
    if (special === 'permission_request') {
      const toolName = entry.metadata?.tool_name as string | undefined;
      return {
        type: 'permission-request',
        machine,
        sessionId,
        ...(toolName !== undefined ? { toolName } : {}),
      };
    }
    if (special === 'ask_question') return { type: 'question', machine, sessionId };
    if (special === 'plan_approval') return { type: 'plan-approval', machine, sessionId };
    // Turn complete — the same marker the transcript hides and the unread dot
    // uses. metadata.stream_end is only ever set on live entries.
    if (!special && entry.metadata?.stream_end) {
      return { type: 'session-finished', machine, sessionId };
    }
    return null;
  }
  if (entry.entryType === 'error') {
    if (special === 'session_died' || special === 'session_failed') {
      return {
        type: 'session-failed',
        machine,
        sessionId,
        ...(entry.content ? { reason: entry.content } : {}),
      };
    }
    return null;
  }
  return null;
}

/**
 * CDX-053 — does this live entry prove the agent is ACTIVELY WORKING the turn?
 * Only such entries may auto-clear a session's unread dot (the old-app "agent
 * working — not waiting on us" nuance). Everything classifyOutputEntry maps to
 * an event marks; everything here clears; the REMAINDER — trailing system
 * status/result/usage lines and error entries — does neither. Pre-fix the
 * clear fired for every unclassified entry, so the trailing system entries a
 * finished turn emits right after its stream_end silently erased the
 * "finished while you were elsewhere" dot before anyone could see it.
 */
export function isAgentActivityEntry(entry: OutputEntry): boolean {
  switch (entry.entryType) {
    case 'text':
    case 'thinking':
    case 'tool_use':
    case 'tool_result':
    case 'progress':
    case 'diff':
      return true;
    default:
      // 'system' (status lines, token/result summaries) and 'error' are turn
      // ARTIFACTS, not activity — they must never eat an attention dot.
      return false;
  }
}

// --- Coordinator: decision + cooldown + formatting → the Notifier port ---

export const NOTIFY_COOLDOWN_MS = 10_000;

export interface NotificationCoordinatorDeps {
  notifier: Notifier;
  /** App visibility — the connection FSM already tracks it (debounced). */
  visible(): boolean;
  /** Master notifications toggle (CDX-048, settings store): when it reads
   *  false, NOTHING fires — neither the OS notification nor the ping chime,
   *  and no cooldown slot is burned. Absent → always enabled. */
  enabled?(): boolean;
  /** In-app attention chime (Web Audio in the platform layer). Deliberately
   *  independent of the notifier: it fires even when the OS notification
   *  permission was never granted. Absent → no chime. */
  ping?(): void;
  /** sessionKeyOf(machine, sessionId) of the session the user is viewing, or
   *  null (no session panel / DM panel in view). Feeds decidePing. */
  activeSessionKey?(): string | null;
  now?(): number;
  cooldownMs?: number;
}

export interface NotificationCoordinator {
  emit(event: NotifyEvent): void;
}

export function createNotificationCoordinator(
  deps: NotificationCoordinatorDeps,
): NotificationCoordinator {
  const now = deps.now ?? Date.now;
  const cooldownMs = deps.cooldownMs ?? NOTIFY_COOLDOWN_MS;
  const recent = new Map<string, number>();

  return {
    emit: (event) => {
      // CDX-048 master toggle: disabled kills BOTH delivery channels at the
      // single decision seam — no notify, no ping, no cooldown burned.
      if (deps.enabled && !deps.enabled()) return;
      const visible = deps.visible();
      const wantNotify = decideNotify(event, visible);
      const wantPing =
        deps.ping !== undefined &&
        decidePing(event, visible, deps.activeSessionKey ? deps.activeSessionKey() : null);
      if (!wantNotify && !wantPing) return;

      // ONE shared cooldown for the ping+notify pair — the live-output path
      // and the heartbeat-transition path (CDX-026b) can both emit for the
      // same card; this is what prevents the double-fire.
      const key = notifyKey(event);
      const at = now();
      const last = recent.get(key);
      if (last !== undefined && at - last < cooldownMs) return;
      recent.set(key, at);
      // Bounded memory: sweep expired keys once the map grows.
      if (recent.size > 64) {
        for (const [k, t] of recent) {
          if (at - t >= cooldownMs) recent.delete(k);
        }
      }

      // Ping BEFORE notify: the chime must fire regardless of what OS
      // delivery does (permission denied, plugin failure).
      if (wantPing) deps.ping?.();
      // The tag lets the platform cancel this delivery later (CDX-026c) when
      // the user opens the session / DM conversation in-app.
      if (wantNotify) deps.notifier.notify({ ...formatNotifyEvent(event), tag: notifyTag(event) });
    },
  };
}

File: crates/client-core/src/notifications.rs
//! Notifications — app-local, store-driven attention events. Port of the pure
//! half of `apps/mobile/src/core/notifications.ts`.
//!
//! NEVER event-sniff: every event is derived from state the phone core already
//! ingested through the typed protocol. The decision is a PURE function of
//! `(event, app visibility)` — a foregrounded app never posts an OS
//! notification (the UI is the notification). Sync catch-up never notifies:
//! only the LIVE output path calls [`classify_output_entry`].

use protocol::common::{EntryBody, NoticeKind, OutputEntry};

// --- event vocabulary ---

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum NotifyEvent {
    PermissionRequest {
        machine: String,
        session_id: String,
        tool_name: Option<String>,
    },
    Question {
        machine: String,
        session_id: String,
    },
    PlanApproval {
        machine: String,
        session_id: String,
    },
    SessionFinished {
        machine: String,
        session_id: String,
    },
    SessionFailed {
        machine: String,
        session_id: String,
        reason: Option<String>,
    },
    DmReceived {
        peer: String,
        peer_label: Option<String>,
        preview: Option<String>,
    },
}

impl NotifyEvent {
    pub fn session(&self) -> Option<(&str, &str)> {
        match self {
            Self::PermissionRequest { machine, session_id, .. }
            | Self::Question { machine, session_id }
            | Self::PlanApproval { machine, session_id }
            | Self::SessionFinished { machine, session_id }
            | Self::SessionFailed { machine, session_id, .. } => Some((machine, session_id)),
            Self::DmReceived { .. } => None,
        }
    }

    /// Stable kind string — the platform Notifier routes delivery on it
    /// (Android: which notification channel).
    pub fn kind_str(&self) -> &'static str {
        match self {
            Self::PermissionRequest { .. } => "permission-request",
            Self::Question { .. } => "question",
            Self::PlanApproval { .. } => "plan-approval",
            Self::SessionFinished { .. } => "session-finished",
            Self::SessionFailed { .. } => "session-failed",
            Self::DmReceived { .. } => "dm-received",
        }
    }
}

/// Human labels resolved by the runtime from its stores at emit time — the
/// event itself only carries machine/session keys. All optional; formatting
/// falls back to the bare kind text when a label is missing.
pub struct NotificationContext<'a> {
    pub session_label: Option<&'a str>,
    pub machine_label: Option<&'a str>,
    /// Display name of the agent the session runs on (from the bridge's agent
    /// catalog), so the text names the agent that actually finished or asked.
    /// Absent (session or catalog not known yet) falls back to "the agent".
    pub agent_label: Option<&'a str>,
}

impl NotificationContext<'_> {
    pub fn none() -> Self {
        Self {
            session_label: None,
            machine_label: None,
            agent_label: None,
        }
    }
}

pub fn session_key_of(machine: &str, session_id: &str) -> String {
    format!("{machine} {session_id}")
}

/// Cooldown / dedup scope: same key + type within the window → one delivery.
pub fn notify_key(event: &NotifyEvent) -> String {
    match event {
        NotifyEvent::DmReceived { peer, .. } => format!("dm {peer}"),
        _ => {
            let (m, s) = event.session().unwrap();
            format!("{} {m} {s}", event.kind_str())
        }
    }
}

/// Parsed back by platform ports (Android turns it into a notification-tap
/// deep link): `:` cannot occur in a machine key (npub/hex) or session id,
/// so ports strip the prefix and split on the LAST colon unambiguously.
pub fn session_notify_tag(machine: &str, session_id: &str) -> String {
    format!("session:{machine}:{session_id}")
}
pub fn dm_notify_tag(peer: &str) -> String {
    format!("dm:{peer}")
}

/// Cancellation scope (CDX-026c) — coarser than [`notify_key`]: opening a
/// session clears EVERY delivered notification for it.
pub fn notify_tag(event: &NotifyEvent) -> String {
    match event {
        NotifyEvent::DmReceived { peer, .. } => dm_notify_tag(peer),
        _ => {
            let (m, s) = event.session().unwrap();
            session_notify_tag(m, s)
        }
    }
}

// --- pure decisions ---

/// A visible app never posts OS notifications; everything here already passed
/// its store-level gate, so `hidden → notify` for every type.
pub fn decide_notify(_event: &NotifyEvent, visible: bool) -> bool {
    !visible
}

/// The in-app chime: ping when hidden, OR when the user is viewing a DIFFERENT
/// session than the event's. A DM event always pings (it already passed the
/// per-conversation unread gate upstream).
pub fn decide_ping(event: &NotifyEvent, visible: bool, active_session_key: Option<&str>) -> bool {
    if !visible {
        return true;
    }
    match event.session() {
        None => true, // dm-received
        Some((m, s)) => Some(session_key_of(m, s).as_str()) != active_session_key,
    }
}

// --- formatting ---

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NotificationContent {
    pub title: String,
    pub body: String,
}

pub fn format_notify_event(event: &NotifyEvent, ctx: &NotificationContext) -> NotificationContent {
    // Session-scoped kinds title "{kind} — {session}" so a multi-session
    // phone can tell cards apart at a glance, and name the machine in the
    // body where it reads naturally. DMs already carry their own labels.
    let titled = |kind: &str| match ctx.session_label {
        Some(s) => format!("{kind} — {s}"),
        None => kind.to_string(),
    };
    let on_machine = |base: &str| match ctx.machine_label {
        Some(m) => format!("{base} on {m}"),
        None => base.to_string(),
    };
    // Who is acting, in body text, and which kind of session, in titles.
    // Sentence-initial and mid-sentence forms of the actor.
    let (agent, agent_mid) = match ctx.agent_label {
        Some(a) => (a, a),
        None => ("The agent", "the agent"),
    };
    let session_kind = |what: &str| match ctx.agent_label {
        Some(a) => format!("{a} session {what}"),
        None => format!("Session {what}"),
    };
    let (title, body) = match event {
        NotifyEvent::PermissionRequest { tool_name, .. } => (
            titled("Permission needed"),
            match tool_name {
                Some(t) => on_machine(&format!("{agent} wants to use {t}")),
                None => on_machine(&format!("{agent} needs permission to proceed")),
            },
        ),
        NotifyEvent::Question { .. } => (
            titled(&format!("Question from {agent_mid}")),
            on_machine(&format!("{agent} is asking you a question")),
        ),
        NotifyEvent::PlanApproval { .. } => (
            titled("Plan ready for review"),
            on_machine("A plan is waiting for your approval"),
        ),
        NotifyEvent::SessionFinished { .. } => (
            titled(&session_kind("finished")),
            on_machine(&format!("{agent} finished the task")),
        ),
        NotifyEvent::SessionFailed { reason, .. } => (
            titled(&session_kind("failed")),
            // Variable text — the machine name would read oddly after it.
            reason
                .clone()
                .unwrap_or_else(|| "The session ended with an error".to_string()),
        ),
        NotifyEvent::DmReceived { peer_label, preview, .. } => (
            peer_label.clone().unwrap_or_else(|| "New message".to_string()),
            preview
                .clone()
                .unwrap_or_else(|| "You received a direct message".to_string()),
        ),
    };
    NotificationContent { title, body }
}

// --- live-output entry classification ---

/// Map ONE live output entry to a notify event, or `None`. The entry kinds
/// that ask for the user, finish a turn, or end a session notify; everything
/// else is ordinary transcript.
pub fn classify_output_entry(
    machine: &str,
    session_id: &str,
    entry: &OutputEntry,
) -> Option<NotifyEvent> {
    let machine = machine.to_string();
    let session_id = session_id.to_string();
    match &entry.body {
        EntryBody::PermissionRequest { tool_name, .. } => Some(NotifyEvent::PermissionRequest {
            machine,
            session_id,
            tool_name: (!tool_name.is_empty()).then(|| tool_name.clone()),
        }),
        // One notification per ask, not per question of a multi-question ask.
        EntryBody::Question { index: 0, .. } => Some(NotifyEvent::Question { machine, session_id }),
        EntryBody::PlanApproval { .. } => Some(NotifyEvent::PlanApproval { machine, session_id }),
        EntryBody::TurnComplete {} => Some(NotifyEvent::SessionFinished { machine, session_id }),
        EntryBody::Notice {
            kind: NoticeKind::SessionDied | NoticeKind::SessionFailed,
            text,
        } => Some(NotifyEvent::SessionFailed {
            machine,
            session_id,
            reason: (!text.is_empty()).then(|| text.clone()),
        }),
        _ => None,
    }
}

/// CDX-053: does this live entry prove the agent is ACTIVELY WORKING? Only such
/// entries may auto-clear a session's unread dot. Status lines, errors,
/// notices and the cards that wait on the user are turn ARTIFACTS, not
/// activity.
pub fn is_agent_activity_entry(entry: &OutputEntry) -> bool {
    matches!(
        entry.body,
        EntryBody::Text { .. }
            | EntryBody::Plan { .. }
            | EntryBody::Thinking { .. }
            | EntryBody::ToolCall { .. }
            | EntryBody::ToolResult { .. }
            | EntryBody::Diff { .. }
    )
}

// --- coordinator ---

pub const NOTIFY_COOLDOWN_MS: u64 = 10_000;

/// What `emit` asks the runtime to do.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum NotifyEffect {
    /// Fire the in-app chime.
    Ping,
    /// Post an OS notification, filed under `tag` (CDX-026c cancellation scope).
    /// `kind` is `NotifyEvent::kind_str` — platform Notifiers route delivery
    /// (Android notification channels) on it.
    Notify {
        content: NotificationContent,
        tag: String,
        kind: String,
    },
}

/// Decision + shared cooldown for the ping/notify pair. The `visible` /
/// `enabled` / `active_session_key` / `now` inputs come from the runtime; the
/// `Notifier` / ping seams are the emitted effects.
#[derive(Debug, Clone)]
pub struct NotificationCoordinator {
    recent: std::collections::BTreeMap<String, u64>,
    cooldown_ms: u64,
}

impl Default for NotificationCoordinator {
    fn default() -> Self {
        Self {
            recent: std::collections::BTreeMap::new(),
            cooldown_ms: NOTIFY_COOLDOWN_MS,
        }
    }
}

/// Everything [`NotificationCoordinator::emit`] needs besides the event
/// itself — grouped so call sites name their booleans instead of listing
/// eight positional arguments.
pub struct EmitInputs<'a> {
    pub visible: bool,
    /// `enabled` false (CDX-048 master toggle) kills BOTH channels at the
    /// single seam — no notify, no ping, no cooldown slot burned.
    pub enabled: bool,
    /// Mirrors the TS `deps.ping !== undefined` (a build with no chime seam
    /// never wants a ping).
    pub ping_available: bool,
    pub active_session_key: Option<&'a str>,
    pub context: NotificationContext<'a>,
    pub now: u64,
}

impl NotificationCoordinator {
    pub fn with_cooldown(cooldown_ms: u64) -> Self {
        Self {
            cooldown_ms,
            ..Self::default()
        }
    }

    pub fn emit(&mut self, event: &NotifyEvent, inputs: EmitInputs<'_>) -> Vec<NotifyEffect> {
        let EmitInputs {
            visible,
            enabled,
            ping_available,
            active_session_key,
            context,
            now,
        } = inputs;
        if !enabled {
            return vec![];
        }
        let want_notify = decide_notify(event, visible);
        let want_ping = ping_available && decide_ping(event, visible, active_session_key);
        if !want_notify && !want_ping {
            return vec![];
        }

        let key = notify_key(event);
        if let Some(&last) = self.recent.get(&key) {
            if now.saturating_sub(last) < self.cooldown_ms {
                return vec![];
            }
        }
        self.recent.insert(key, now);
        if self.recent.len() > 64 {
            self.recent
                .retain(|_, &mut t| now.saturating_sub(t) < self.cooldown_ms);
        }

        let mut effects = Vec::new();
        // Ping BEFORE notify — the chime must fire regardless of OS delivery.
        if want_ping {
            effects.push(NotifyEffect::Ping);
        }
        if want_notify {
            effects.push(NotifyEffect::Notify {
                content: format_notify_event(event, &context),
                tag: notify_tag(event),
                kind: event.kind_str().to_string(),
            });
        }
        effects
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    /// Common EmitInputs for the coordinator tests — `EmitInputs { field: x,
    /// ..inputs(...) }` overrides single fields at the call site.
    fn inputs(visible: bool, enabled: bool, ping_available: bool) -> EmitInputs<'static> {
        EmitInputs {
            visible,
            enabled,
            ping_available,
            active_session_key: None,
            context: NotificationContext::none(),
            now: 0,
        }
    }

    fn perm(m: &str, s: &str) -> NotifyEvent {
        NotifyEvent::PermissionRequest {
            machine: m.into(),
            session_id: s.into(),
            tool_name: None,
        }
    }
    fn dm(peer: &str) -> NotifyEvent {
        NotifyEvent::DmReceived {
            peer: peer.into(),
            peer_label: None,
            preview: None,
        }
    }
    fn entry(body: EntryBody) -> OutputEntry {
        OutputEntry::new("1970-01-01T00:00:00.000Z", body)
    }

    #[test]
    fn decide_notify_is_just_hidden() {
        assert!(decide_notify(&perm("m", "s"), false));
        assert!(!decide_notify(&perm("m", "s"), true));
        assert!(decide_notify(&dm("p"), false));
    }

    #[test]
    fn decide_ping_matrix() {
        // hidden -> always
        assert!(decide_ping(&perm("m", "s"), false, Some("m s")));
        // visible + viewing this session -> no
        assert!(!decide_ping(&perm("m", "s"), true, Some("m s")));
        // visible + viewing a different session -> yes
        assert!(decide_ping(&perm("m", "s"), true, Some("m other")));
        // visible + no session panel -> yes
        assert!(decide_ping(&perm("m", "s"), true, None));
        // dm -> always pings when it reaches here
        assert!(decide_ping(&dm("p"), true, Some("m s")));
    }

    #[test]
    fn keys_and_tags() {
        assert_eq!(notify_key(&perm("m", "s")), "permission-request m s");
        assert_eq!(notify_key(&dm("p")), "dm p");
        assert_eq!(notify_tag(&perm("m", "s")), "session:m:s");
        assert_eq!(notify_tag(&dm("p")), "dm:p");
    }

    #[test]
    fn formatting_covers_every_variant() {
        assert_eq!(format_notify_event(&perm("m", "s"), &NotificationContext::none()).title, "Permission needed");
        assert_eq!(
            format_notify_event(&NotifyEvent::PermissionRequest {
                machine: "m".into(),
                session_id: "s".into(),
                tool_name: Some("Bash".into())
            }, &NotificationContext::none())
            .body,
            "The agent wants to use Bash"
        );
        assert_eq!(
            format_notify_event(&NotifyEvent::SessionFailed {
                machine: "m".into(),
                session_id: "s".into(),
                reason: Some("boom".into())
            }, &NotificationContext::none())
            .body,
            "boom"
        );
        assert_eq!(
            format_notify_event(&NotifyEvent::DmReceived {
                peer: "p".into(),
                peer_label: Some("Alice".into()),
                preview: Some("hi".into())
            }, &NotificationContext::none()),
            NotificationContent { title: "Alice".into(), body: "hi".into() }
        );
    }

    #[test]
    fn formatting_uses_the_runtime_resolved_labels() {
        let ctx = NotificationContext {
            session_label: Some("refactor-api"),
            machine_label: Some("laptop-01"),
            agent_label: Some("Claude Code"),
        };
        let content = format_notify_event(&perm("m", "s"), &ctx);
        assert_eq!(content.title, "Permission needed — refactor-api");
        assert_eq!(content.body, "Claude Code needs permission to proceed on laptop-01");
        // No labels yet (unknown keys) — bare kind text, unchanged.
        let bare = format_notify_event(&perm("m", "s"), &NotificationContext::none());
        assert_eq!(bare.title, "Permission needed");
        assert_eq!(bare.body, "The agent needs permission to proceed");
    }

    #[test]
    fn formatting_names_the_sessions_agent() {
        let finished = NotifyEvent::SessionFinished { machine: "m".into(), session_id: "s".into() };
        let question = NotifyEvent::Question { machine: "m".into(), session_id: "s".into() };
        let failed = NotifyEvent::SessionFailed { machine: "m".into(), session_id: "s".into(), reason: None };
        let ctx = |agent| NotificationContext {
            session_label: Some("fix-ci"),
            machine_label: Some("laptop-01"),
            agent_label: Some(agent),
        };

        let open_code = ctx("OpenCode");
        assert_eq!(
            format_notify_event(&finished, &open_code),
            NotificationContent {
                title: "OpenCode session finished — fix-ci".into(),
                body: "OpenCode finished the task on laptop-01".into(),
            }
        );
        assert_eq!(format_notify_event(&question, &open_code).title, "Question from OpenCode — fix-ci");
        assert_eq!(format_notify_event(&failed, &open_code).title, "OpenCode session failed — fix-ci");

        let claude_code = ctx("Claude Code");
        assert_eq!(
            format_notify_event(&finished, &claude_code),
            NotificationContent {
                title: "Claude Code session finished — fix-ci".into(),
                body: "Claude Code finished the task on laptop-01".into(),
            }
        );

        // Session not in the store yet — the generic wording.
        let bare = format_notify_event(&finished, &NotificationContext::none());
        assert_eq!(bare.title, "Session finished");
        assert_eq!(bare.body, "The agent finished the task");
        assert_eq!(
            format_notify_event(&question, &NotificationContext::none()).title,
            "Question from the agent"
        );
    }

    fn body(v: serde_json::Value) -> EntryBody {
        serde_json::from_value(v).unwrap()
    }

    #[test]
    fn classify_output_entry_maps_the_cards_that_want_the_user() {
        let c = |b| classify_output_entry("m", "s", &entry(body(b)));
        assert_eq!(
            c(json!({"entryType":"permission_request","requestId":"r","toolName":"Edit","kind":"edit","title":"a.rs","options":[]})),
            Some(NotifyEvent::PermissionRequest { machine: "m".into(), session_id: "s".into(), tool_name: Some("Edit".into()) })
        );
        assert_eq!(
            c(json!({"entryType":"question","requestId":"q","index":0,"count":2,"question":"?"})),
            Some(NotifyEvent::Question { machine: "m".into(), session_id: "s".into() })
        );
        // later questions of the same ask do not notify again
        assert_eq!(c(json!({"entryType":"question","requestId":"q","index":1,"count":2,"question":"?"})), None);
        assert_eq!(
            c(json!({"entryType":"plan_approval","requestId":"p","options":[]})),
            Some(NotifyEvent::PlanApproval { machine: "m".into(), session_id: "s".into() })
        );
        assert_eq!(
            c(json!({"entryType":"turn_complete"})),
            Some(NotifyEvent::SessionFinished { machine: "m".into(), session_id: "s".into() })
        );
        assert_eq!(
            c(json!({"entryType":"notice","kind":"session_died","text":"it died"})),
            Some(NotifyEvent::SessionFailed { machine: "m".into(), session_id: "s".into(), reason: Some("it died".into()) })
        );
        // no-ops
        assert_eq!(c(json!({"entryType":"notice","kind":"session_restart","text":"x"})), None);
        assert_eq!(c(json!({"entryType":"status","text":"compacting"})), None);
        assert_eq!(c(json!({"entryType":"text","role":"agent","text":"hello"})), None);
        assert_eq!(c(json!({"entryType":"error","text":"boom"})), None);
    }

    #[test]
    fn agent_activity_excludes_artifacts() {
        for b in [
            json!({"entryType":"text","role":"agent","text":"x"}),
            json!({"entryType":"plan","text":"x"}),
            json!({"entryType":"thinking","text":"x"}),
            json!({"entryType":"tool_call","callId":"c","toolName":"Read","kind":"read","title":"a"}),
            json!({"entryType":"tool_result","callId":"c","text":"x"}),
            json!({"entryType":"diff","path":"a","lines":[]}),
        ] {
            assert!(is_agent_activity_entry(&entry(body(b.clone()))), "{b}");
        }
        for b in [
            json!({"entryType":"status","text":"x"}),
            json!({"entryType":"error","text":"x"}),
            json!({"entryType":"turn_complete"}),
            json!({"entryType":"resolved","requestId":"r","summary":"ok"}),
            json!({"entryType":"permission_request","requestId":"r","toolName":"Edit","kind":"edit","title":"a","options":[]}),
        ] {
            assert!(!is_agent_activity_entry(&entry(body(b.clone()))), "{b}");
        }
    }

    // --- coordinator ---

    #[test]
    fn master_toggle_off_kills_both_channels_and_burns_no_cooldown() {
        let mut co = NotificationCoordinator::default();
        assert!(co.emit(&perm("m", "s"), inputs(false, false, true)).is_empty());
        // enabled again immediately: not blocked by a cooldown slot
        assert!(!co.emit(&perm("m", "s"), EmitInputs { now: 1, ..inputs(false, true, true) }).is_empty());
    }

    #[test]
    fn a_visible_user_watching_the_session_gets_nothing() {
        let mut co = NotificationCoordinator::default();
        assert!(co.emit(&perm("m", "s"), EmitInputs { visible: true, active_session_key: Some("m s"), ..inputs(true, true, true) }).is_empty());
    }

    #[test]
    fn ping_comes_before_notify_and_shares_one_cooldown() {
        let mut co = NotificationCoordinator::default();
        let eff = co.emit(&perm("m", "s"), inputs(false, true, true));
        assert_eq!(eff.len(), 2);
        assert_eq!(eff[0], NotifyEffect::Ping);
        assert!(matches!(eff[1], NotifyEffect::Notify { .. }));

        // same key inside the window: nothing (the live path AND the heartbeat
        // path can both emit for one card — this is the double-fire guard).
        assert!(co.emit(&perm("m", "s"), EmitInputs { now: NOTIFY_COOLDOWN_MS - 1, ..inputs(false, true, true) }).is_empty());
        // window elapsed: fires again
        assert_eq!(co.emit(&perm("m", "s"), EmitInputs { now: NOTIFY_COOLDOWN_MS, ..inputs(false, true, true) }).len(), 2);
    }

    #[test]
    fn ping_only_when_the_os_notification_is_suppressed_but_the_user_is_elsewhere() {
        let mut co = NotificationCoordinator::default();
        let eff = co.emit(&perm("m", "s"), EmitInputs { active_session_key: Some("m other"), ..inputs(true, true, true) });
        assert_eq!(eff, vec![NotifyEffect::Ping]);
    }

    #[test]
    fn no_ping_seam_means_no_ping_effect() {
        let mut co = NotificationCoordinator::default();
        let eff = co.emit(&perm("m", "s"), inputs(false, true, false));
        assert_eq!(eff.len(), 1);
        assert!(matches!(eff[0], NotifyEffect::Notify { .. }));
    }
}
